File: companies/facebook/p632/Solution.java
class Solution {
   class Node{
       int group;
       int value;
       public Node(int group, int value){
           this.group = group;
           this.value = value;
       }
   }
   public int[] smallestRange(List<List<Integer>> nums) {
       if (nums == null || nums.size() == 0){
           return new int[0];
       }

       int k = nums.size();
       List<Node> list = new ArrayList<>();
       for (int i = 0; i < nums.size(); i++){
           for (int j = 0; j < nums.get(i).size(); j++){
               list.add(new Node(i, nums.get(i).get(j)));
           }
       }

       Collections.sort(list, (Node a, Node b) -> (a.value - b.value));

       Map<Integer, Integer> map = new HashMap<>();
       int diff = Integer.MAX_VALUE;
       int j = 0;
       int[] res = new int[2];

       for (int i = 0; i < list.size(); i++){
           while (j < list.size() && map.size() < k){
               Node tmp = list.get(j);
               if (!map.containsKey(tmp.group)){
                   map.put(tmp.group, 0);
               }
               map.put(tmp.group, map.get(tmp.group) + 1);
               j ++;
           }

           if (map.size() == k){
               if (list.get(j - 1).value - list.get(i).value < diff){
                   diff = list.get(j - 1).value - list.get(i).value;
                   res[0] = list.get(i).value;
                   res[1] = list.get(j - 1).value;
               }
           }

           if (map.get(list.get(i).group) == 1){
               map.remove(list.get(i).group);
           }
           else{
               map.put(list.get(i).group, map.get(list.get(i).group) - 1);
           }
       }

       return res;

   }
}

File: companies/facebook/p632/Solution.py
import sys


class Solution:
    def smallestRange(self, nums):
        """
        :type nums: List[List[int]]
        :rtype: List[int]
        """
        if not nums or len(nums) == 0:
            return []

        flat_nums = []
        for i in range(len(nums)):
            for num in nums[i]:
                flat_nums.append((num, i))

        flat_nums.sort()

        diff = sys.maxsize
        answer = [0, 0]
        right = 0
        table = dict()
        for left in range(len(flat_nums)):
            while right < len(flat_nums) and not self.valid(table, nums):
                group = flat_nums[right][1]
                if group not in table:
                    table[group] = 1
                else:
                    table[group] += 1

                right += 1

            if self.valid(table, nums):
                curDiff = flat_nums[right-1][0] - flat_nums[left][0]
                if curDiff < diff:
                    diff = curDiff
                    answer = [flat_nums[left][0], flat_nums[right-1][0]]

            left_group = flat_nums[left][1]
            if table[left_group] - 1 == 0:
                del table[left_group]
            else:
                table[left_group] -= 1

        return answer

    def valid(self, table, nums):
        return len(table) == len(nums)
